File: src/main/java/gg/generations/rarecandy/tools/gui/FullscreenPass.java
package gg.generations.rarecandy.tools.gui;

import gg.generations.rarecandy.pokeutils.reader.ITextureLoader;
import gg.generations.rarecandy.renderer.pipeline.traditional.TraditionalPipeline;
import gg.generations.rarecandy.renderer.pipeline.util.Scope;
import gg.generations.rarecandy.renderer.textures.FrameBuffer;
import org.lwjgl.opengl.GL11C;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import java.util.function.Supplier;

public final class FullscreenPass {
    public record Input(String uniform, Source source) {}

    private final Supplier<TraditionalPipeline.Builder> base;
    private final List<Input> inputs;
    private final Consumer<TraditionalPipeline.Builder> extraUniforms;
    private final BooleanSupplier enabled;

    // Set by the chain immediately before draw; the sampler suppliers read them.
    private FrameBuffer gbuffer;
    private FrameBuffer previous;

    private TraditionalPipeline pipeline;

    private FullscreenPass(Builder builder) {
        this.base = builder.base;
        this.inputs = List.copyOf(builder.inputs);
        this.extraUniforms = builder.extraUniforms;
        this.enabled = builder.enabled;
        this.pipeline = compile();
    }

    public static Builder of(Supplier<TraditionalPipeline.Builder> base) {
        return new Builder(base);
    }

    public boolean isEnabled() {
        return enabled.getAsBoolean();
    }

    public void reload() {
        pipeline = compile();
    }

    private TraditionalPipeline compile() {
        TraditionalPipeline.Builder builder = base.get();

        // Unit is the declaration index, so no two inputs can collide.
        for (int unit = 0; unit < inputs.size(); unit++) {
            Input input = inputs.get(unit);
            builder.autoSampler2D(Scope.GLOBAL, input.uniform(), unit, ctx -> resolve(input.source()));
        }

        extraUniforms.accept(builder);
        return builder.build();
    }

    private int resolve(Source source) {
        return switch (source) {
            case Source.Attachment a -> gbuffer.getColorAttachment(a.index()).id();
            case Source.Depth d -> gbuffer.getDepthTexture().id();
            case Source.Previous p -> previous.getColorAttachment(0).id();
            case Source.Named n -> ITextureLoader.instance().getTexture(n.key()).id();
        };
    }

    void render(FrameBuffer gbuffer, FrameBuffer previous, FrameBuffer target) {
        this.gbuffer = gbuffer;
        this.previous = previous;

        target.bindAndSetViewport();
        target.setDrawAll();
        draw();
    }

    void renderToScreen(FrameBuffer gbuffer, FrameBuffer previous, int width, int height) {
        this.gbuffer = gbuffer;
        this.previous = previous;

        FrameBuffer.unbindDraw();
        GL11C.glViewport(0, 0, width, height);
        draw();
    }

    private void draw() {
        pipeline.useProgram();
        pipeline.bindGlobal();
        GL11C.glDrawArrays(GL11C.GL_TRIANGLES, 0, 3);
    }

    public static final class Builder {
        private final Supplier<TraditionalPipeline.Builder> base;
        private final List<Input> inputs = new ArrayList<>();
        private Consumer<TraditionalPipeline.Builder> extraUniforms = b -> {};
        private BooleanSupplier enabled = () -> true;

        private Builder(Supplier<TraditionalPipeline.Builder> base) {
            this.base = base;
        }

        public Builder reads(String uniform, Source source) {
            inputs.add(new Input(uniform, source));
            return this;
        }

        public Builder uniforms(Consumer<TraditionalPipeline.Builder> consumer) {
            this.extraUniforms = consumer;
            return this;
        }

        public Builder enabledWhen(BooleanSupplier enabled) {
            this.enabled = enabled;
            return this;
        }

        public FullscreenPass build() {
            return new FullscreenPass(this);
        }
    }
}

File: src/main/java/gg/generations/rarecandy/tools/gui/FullscreenPass.kt
package gg.generations.rarecandy.tools.gui

import gg.generations.rarecandy.pokeutils.reader.ITextureLoader
import gg.generations.rarecandy.renderer.pipeline.traditional.TraditionalPipeline
import gg.generations.rarecandy.renderer.pipeline.util.Scope
import gg.generations.rarecandy.renderer.textures.framebuffer.FrameBuffer
import gg.generations.rarecandy.renderer.textures.framebuffer.TextureTarget
import org.lwjgl.opengl.GL11C

data class FullscreenPass(
    private val base: () -> TraditionalPipeline.Builder,
    private val inputs: List<Input>,
    private val extraUniforms: (TraditionalPipeline.Builder) -> Unit,
    val enabled: () -> Boolean,
    val output: String?,
    val scale: Float
) {
    data class Input(val uniform: String, val source: Source)

    constructor(builder: Builder) : this(
        builder.base,
        builder.inputs,
        builder.extraUniforms,
        builder.enabled,
        builder.output,
        builder.scale)

    private var namedTargets: Map<String, FrameBuffer> = mapOf<String, FrameBuffer>()
    private var pipeline: TraditionalPipeline?
    private lateinit var gbuffer: FrameBuffer
    private lateinit var previous: FrameBuffer

    init {
        pipeline = compile()
    }

    fun reload() {
        var compiled = compile()
        pipeline?.destroy()
        pipeline = compiled
    }

    fun destroy() {
        pipeline?.destroy()
        pipeline = null
    }

    private fun compile(): TraditionalPipeline {
        val builder = base.invoke()

        // Unit is the declaration index, so no two inputs can collide.
        inputs.forEachIndexed { unit, input ->
            builder.autoSampler2D(Scope.GLOBAL, input.uniform, unit) { ctx -> resolve(input.source) }
        }

        extraUniforms.invoke(builder)
        return builder.build()
    }

    private fun resolve(source: Source): Int {
        return when (source) {
            is Source.Attachment -> gbuffer.getColorAttachment(source.index)?.id ?: 0
            is Source.Depth -> gbuffer.getDepthTexture().id
            is Source.Previous -> previous.getColorAttachment(0)?.id ?: 0
            is Source.Named -> ITextureLoader.instance().getTexture(source.key).id
            is Source.Pass -> {
                val buffer = namedTargets.get(source.key) ?: throw IllegalStateException("Pass reads '${source.key}' but no earlier enabled pass writes it")
                buffer.getColorAttachment(0)?.id ?: 0
            }

            else -> 0
        }
    }

    fun render(gbuffer: FrameBuffer, previous: FrameBuffer, named: Map<String, FrameBuffer>, target: FrameBuffer) {
        this.gbuffer = gbuffer
        this.previous = previous
        this.namedTargets = named

        target.bindAndSetViewport()
        target.setDrawAll()
        draw()
    }

    fun renderToScreen(gBuffer: FrameBuffer, previous: FrameBuffer, named: Map<String, FrameBuffer>, width: Int, height: Int) {
        this.previous = previous
        this.namedTargets = named

        FrameBuffer.unbindDraw()
        GL11C.glViewport(0, 0, width, height)
        draw()
    }

    private fun draw() {
        if(pipeline == null) return
        GL11C.glDisable(GL11C.GL_DEPTH_TEST)
        GL11C.glDisable(GL11C.GL_BLEND)


        pipeline!!.useProgram()
        pipeline!!.bindGlobal()
        GL11C.glDrawArrays(GL11C.GL_TRIANGLES, 0, 3)
    }

    class Builder(internal val base: () -> TraditionalPipeline.Builder) {
        internal val inputs = mutableListOf<Input>()
        internal var extraUniforms: (TraditionalPipeline.Builder) -> Unit = { }
        internal var enabled: () -> Boolean = { true }
        internal var output: String? = null
        internal var scale = 1.0f

        fun reads(uniform: String, source: Source): Builder {
            inputs.add(Input(uniform, source))
            return this
        }

        fun uniforms(consumer: (TraditionalPipeline.Builder) -> Unit): Builder {
            this.extraUniforms = consumer
            return this
        }

        fun enabledWhen(enabled: () -> Boolean): Builder {
            this.enabled = enabled
            return this
        }

        fun writesTo(key: String): Builder {
            this.output = key
            return this
        }

        fun scale(scale: Float): Builder {
            if (scale < 0.0f) throw IllegalArgumentException("Scale can't be less than 0, it was " + scale + "instead")
            this.scale = scale
            return this
        }

        fun build(): FullscreenPass = FullscreenPass(this)
    }

    companion object {
        fun of(base: () -> TraditionalPipeline.Builder): Builder = Builder(base)
    }
}